providers/implementations/kem/hybrid_kem.c

```diff
@@ -0,0 +1,260 @@
+        return 0;
+
+    for (i = 0; i < ctx->info->num_algs; i++) {
+        EVP_PKEY_CTX_free(ctx->ctxs[i]);
+        ctx->ctxs[i] = EVP_PKEY_CTX_new_from_pkey(ctx->libctx,
+                                                  key->keys[i], key->propq);
+        if (ctx->ctxs[i] == NULL || !(*initf)(ctx->ctxs[i], auth->keys[i], params))
+            return 0;
+    }
+    return ossl_hybrid_set_ctx_params(ctx, params);
+}
+
+int ossl_hybrid_kem_auth_encapsulate_init(void *vctx, void *vkey,
+                                          void *vauthpriv,
+                                          const OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    HYBRID_PKEY *key = (HYBRID_PKEY *)vkey;
+    HYBRID_PKEY *authpriv = (HYBRID_PKEY *)vauthpriv;
+
+    return hybrid_kem_auth_init(ctx, key, authpriv, params,
+                                &EVP_PKEY_auth_encapsulate_init);
+}
+
+int ossl_hybrid_kem_auth_decapsulate_init(void *vctx, void *vkey,
+                                          void *vauthpub,
+                                          const OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    HYBRID_PKEY *key = (HYBRID_PKEY *)vkey;
+    HYBRID_PKEY *authpub = (HYBRID_PKEY *)vauthpub;
+
+    return hybrid_kem_auth_init(ctx, key, authpub, params,
+                                &EVP_PKEY_auth_decapsulate_init);
+}
+
+int ossl_hybrid_kem_encapsulate(void *vctx, unsigned char *enc, size_t *enclen,
+                                unsigned char *secret, size_t *secretlen)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    size_t a, b;
+    unsigned int i;
+
+    if (ctx->key_length == 0)
+        return 0;
+
+    if (enc == NULL) {
+        if (enclen != NULL)
+            *enclen = ctx->ciphertext_bytes;
+        if (secretlen != NULL)
+            *secretlen = ctx->shared_secret_bytes;
+        return enclen != NULL || secretlen != NULL;
+    }
+
+    /* Error checking */
+    if (*secretlen < ctx->shared_secret_bytes) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH, "*secretlen too small");
+        return 0;
+    }
+    if (*enclen < ctx->ciphertext_bytes) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH, "*enclen too small");
+        return 0;
+    }
+
+    /* Set outputs */
+    *enclen = ctx->ciphertext_bytes;
+    *secretlen = ctx->shared_secret_bytes;
+
+    /* Encapsulate */
+    for (i = 0; i < ctx->info->num_algs; i++) {
+        a = ctx->info->alg[i].ciphertext_bytes;
+        b = ctx->info->alg[i].shared_secret_bytes;
+        if (!EVP_PKEY_encapsulate(ctx->ctxs[i], enc, &a, secret, &b))
+            return 0;
+        if (a != ctx->info->alg[i].ciphertext_bytes
+                || b != ctx->info->alg[i].shared_secret_bytes) {
+            ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH,
+                           "secretlen or enclen changed");
+            return 0;
+        }
+        secret += ctx->info->alg[i].shared_secret_bytes;
```

Comment:
If I get this right, the order of algorithms in `info->alg` defines the order of concatenations (which is OK). However, does this then require a change in logic as to how the common name is created? If I don't get it wrong, it's currently the concatenation of constituent names in that order, right? However, according to https://www.ietf.org/archive/id/draft-kwiatkowski-tls-ecdhe-mlkem-02.html, section 3.1.3, "X25519MLKEM768" then has to be declared "ML-KEM-768-X25519" (in that order) for the logic here to work correctly. The order of constituent algorithms and names for "SecP256r1MLKEM768" is unconditionally correct.

Reply:
Not quite.  The order of the concatenation is defined by `info->alg` but the overall name can be anything.  That's defined separately in `providers/defltprov.c`.